includes/rest-api/class-sensei-rest-api-data-port-base.php

```diff
@@ -0,0 +1,199 @@
+
+	/**
+	 * Routes namespace.
+	 *
+	 * @var string
+	 */
+	protected $namespace;
+
+	/**
+	 * Sensei_REST_API_Data_Port_Base constructor.
+	 *
+	 * @param string $namespace Routes namespace.
+	 */
+	public function __construct( $namespace ) {
+		$this->namespace = $namespace;
+	}
+
+	/**
+	 * Register the REST API endpoints for the Importer.
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			$this->rest_base,
+			[
+				[
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => [ $this, 'request_get_job' ],
+					'permission_callback' => [ $this, 'can_user_access_rest_api' ],
+				],
+				[
+					'methods'             => WP_REST_Server::CREATABLE,
+					'callback'            => [ $this, 'request_post_job' ],
+					'permission_callback' => [ $this, 'can_user_access_rest_api' ],
+				],
+				[
+					'methods'             => WP_REST_Server::DELETABLE,
+					'callback'            => [ $this, 'request_delete_job' ],
+					'permission_callback' => [ $this, 'can_user_access_rest_api' ],
+				],
+				'schema' => [ $this, 'get_item_schema' ],
+			]
+		);
+	}
+
+	/**
+	 * Get the current import job.
+	 *
+	 * @return WP_REST_Response|WP_Error
+	 */
+	public function request_get_job() {
+		$job = $this->get_active_job();
+		if ( ! $job ) {
+			return new WP_Error(
+				'rest_no_active_job',
+				__( 'No job has been created.', 'sensei-lms' ),
+				array( 'status' => 404 )
+			);
+		}
+
+		$response = new WP_REST_Response();
+		$response->set_data( $this->prepare_to_serve_job( $job ) );
+
+		return $response;
+	}
+
+	/**
+	 * Initialize an import job.
+	 *
+	 * @return WP_REST_Response
+	 */
+	public function request_post_job() {
+		$job     = $this->get_active_job();
+		$created = false;
+		if ( ! $job ) {
+			$job     = Sensei_Data_Port_Manager::instance()->create_import_job( get_current_user_id() );
+			$created = true;
+		}
+
+		$response = new WP_REST_Response();
+		$response->set_data( $this->prepare_to_serve_job( $job ) );
+		if ( $created ) {
+			$response->set_status( 201 );
+		}
+
+		return $response;
+	}
+
+	/**
+	 * Cancel the currently active job.
+	 *
+	 * @return WP_REST_Response|WP_Error
+	 */
+	public function request_delete_job() {
+		$job = $this->get_active_job();
+		if ( ! $job ) {
+			return new WP_Error(
+				'rest_no_active_job',
+				__( 'No job has been created.', 'sensei-lms' ),
+				array( 'status' => 404 )
+			);
+		}
+
+		Sensei_Data_Port_Manager::instance()->cancel_job( $job->get_job_id() );
+
+		$response = new WP_REST_Response();
+		$response->set_data(
+			[
+				'deleted'  => true,
+				'previous' => $this->prepare_to_serve_job( $job ),
+			]
+		);
+
+		return $response;
+	}
+
+	/**
+	 * Get the active job for this user.
+	 *
+	 * @return Sensei_Data_Port_Job|null
+	 */
+	protected function get_active_job() {
+		return Sensei_Data_Port_Manager::instance()->get_active_job( $this->handler_class, get_current_user_id() );
```

Comment:
Maybe hide `$this->handler_class` behind an abstract getter?

Reply:
Good idea 9583a9ae99f9a2b544d68ec8bfffe6f8d8d12b03